### PR TITLE
Introduce CustomActions feature (closes #1535)

### LINKED ISSUE #1535: Provide an ability to add custom actions to the test controller
Based on discussions: 
https://testcafe-discuss.devexpress.com/t/use-custom-actions-on-test-controller/228
https://testcafe-discuss.devexpress.com/t/how-to-chain-pagemodel-methods/313

### Are you requesting a feature or reporting a bug?

Feature

### What is the expected behavior?

Provide an ability to add custom actions to the test controller or chain methods that contain a test controller chain

### PR DESCRIPTION
**Pull request by Artem-Babich**
## Purpose
In some cases, you need to create custom actions that contain multiple default actions or other custom actions.

## Approach
Provide the ability to describe and register custom actions in a JS configuration file. Create a **TestController.customActions** property containing these actions. If the user function returns a value, the result of calling the user action will be that value. If the function does not return a value, the result will be a TestController object, which you can chain further.

### Define Custom Actions

```js
// JS Configuration file

module.exports = {
    customActions: {
        async makeSomething (selector) {
           await this.click(selector);
        },
        async getSelectorValue (selector) {
         return await Selector(selector).innerText;
       },
    }
}
```

### Use Custom Actions
```js
test('Check span value', async t => {
    const spanValue = await t.customActions.getSelectorValue('#result');

    await t.expect(spanValue).eql('OK');
});


test('Click the button and check span value', async t => {
    const spanValue = await t.customActions.makeSomething()
                             .customActions.getSelectorValue('#result');

    await t.expect(spanValue).eql('OK');
});

```

### Reporter Changes (For Dashboard team):

- runCustomAction is a command that fires **reportTestActionStart** **before** any inner action starts and fires **reportTestActionDone** **after** the latest inner action fineshed.
- you can identify each custom action run using **command.actionId** property.
- in the case of concurrency mode you can also use testRunId.
- you can also access the result of the runCustomAction command(the value returned by the custom action) in the **reportTestActionDone** function using **command.actionResult** property.

An example of reporter:

```js
const customActionsStack = {}
const shouldReportInnerActions = false;

function isCustomAction (name /*or type */) {
   return name ==='runCustomAction';
   // or return type === 'run-custom-action';
}

{
  reportTestActionStart: (actionName, { command, testRunId }) => {
     if(isCustomAction(actionName /* or command.type */ ) ) {
        const { actionId, name } = command;
        customActionsStack[testRunId].push({ actionId, name }) ;
     }
  },
  reportTestActionDone:  (name, { command, testRunId }) => {
     if( isCustomAction(actionName /* or command.type */ ) ) {
        customActionsStack[testRunId].pop();
  
        const { actionResult, actionId, name } = command;
        // Do something with actionResult
        return;
     }
  
     if (!shouldReportInnerActions && customActionsStack[testRunId].length) {
        // Do not report action
     } else {
        // Report action
     }   
  }
}
```

## References
Closes #1535 
## Pre-Merge TODO
- [x] Write tests for your proposed changes
- [x] Make sure that existing tests do not fail
- [x] Make sure that the documentation link is correct in the Error template

### CODE REVIEW

**Review by AndreyBelym — APPROVED** [review]

File: test/server/runner-test.js [hunk 1]
```diff
@@ -1336,6 +1336,41 @@ describe('Runner', () => {
                                         'The testRun.globalAfter hook (string) is not of expected type (function).');
             }
         });
+
+        it('Should raise an error if customActions property belongs to invalid type', async function () {
```
Comment (miherlosev) [inline test/server/runner-test.js:1340]: Add a test demonstrating reporting of custom actions.

**Review by miherlosev — CHANGES_REQUESTED** [review]

File: src/reporter/command/command-formatter.ts [hunk 2]
```diff
@@ -2,7 +2,7 @@ import { isEmpty } from 'lodash';
 import { ExecuteSelectorCommand, ExecuteClientFunctionCommand } from '../../test-run/commands/observation';
 import {
     NavigateToCommand,
-    PressKeyCommand,
+    PressKeyCommand, RunCustomActionCommand,
```
Comment (miherlosev) [inline src/reporter/command/command-formatter.ts:5]: Move to the new line

File: src/errors/runtime/templates.js [hunk 3]
```diff
@@ -139,4 +139,6 @@ export default {
     [RUNTIME_ERRORS.invalidSkipJsErrorsOptionsObjectProperty]:       `The "{optionName}" option does not exist. Use the following options to configure skipJsErrors: ${getConcatenatedValuesString(Object.keys(SKIP_JS_ERRORS_OPTIONS_OBJECT_OPTION_NAMES))}.`,
     [RUNTIME_ERRORS.invalidSkipJsErrorsCallbackWithOptionsProperty]: `The "{optionName}" option does not exist. Use the following options to configure skipJsErrors callback: ${getConcatenatedValuesString(Object.keys(SKIP_JS_ERRORS_CALLBACK_WITH_OPTIONS_OPTION_NAMES))}.`,
     [RUNTIME_ERRORS.invalidCommandInJsonCompiler]:                   `TestCafe terminated the test run. The "{path}" file contains an unknown Chrome User Flow action "{action}". Remove the action to continue. Refer to the following article for the definitive list of supported Chrome User Flow actions: https://testcafe.io/documentation/403998/guides/experimental-capabilities/chrome-replay-support#supported-replay-actions`,
+    [RUNTIME_ERRORS.invalidCustomActionsOptionType]:                 `The value of the customActions option does not belong to type Object. Refer to the following article for custom action setup instructions: CUSTOM_ACTIONS_LINK`,
```
Comment (miherlosev) [inline src/errors/runtime/templates.js:142]: It's better adding `DOCUMENTATION_LINKS.CUSTOM_ACTIONS` constant [here](https://github.com/DevExpress/testcafe/pull/7393/files#diff-2b693f1a4eee477500a439748a3a10702958774c09ccd8813d36fa8bbffb3eb4L11) and use it in templates.

File: src/reporter/command/command-formatter.ts [hunk 4]
```diff
@@ -125,6 +128,11 @@ export class CommandFormatter {
         return command.url;
     }
 
+    private _assignCustomActionResult (formatedCommand: FormattedCommand) :void {
+        if (this._result !== void 0)
```
Comment (miherlosev) [inline src/reporter/command/command-formatter.ts:132]: I guess, we can always apply the action result even if it equals `undefined`.

**Review by miherlosev — CHANGES_REQUESTED** [review]

File: src/reporter/command/command-formatter.ts [hunk 5]
```diff
@@ -125,6 +129,10 @@ export class CommandFormatter {
         return command.url;
     }
 
+    private _prepareCustomActionResult (): unknown {
```
Comment (miherlosev) [inline src/reporter/command/command-formatter.ts:132]: You can remove this function since it's used only in one place and contains only one instruction.

**Review by miherlosev — CHANGES_REQUESTED** [review]

**Review by miherlosev — APPROVED** [review]

**Pull request merged**